src/view.c

```diff
@@ -18,6 +18,80 @@
 #define LAB_FALLBACK_WIDTH  640
 #define LAB_FALLBACK_HEIGHT 480
 
+static bool
+matches_criteria(struct view *view, enum lab_view_criteria criteria)
+{
+	if (!isfocusable(view)) {
+		return false;
+	}
+	if (criteria & LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		/*
+		 * Always-on-top views are always on the current desktop and are
+		 * special in that they live in a different tree.
+		 */
+		struct server *server = view->server;
+		if (view->scene_tree->node.parent != server->workspace_current->tree
+				&& !view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_ALWAYS_ON_TOP) {
+		if (view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER) {
+		if (window_rules_get_property(view, "skipWindowSwitcher") == LAB_PROP_TRUE) {
+			return false;
+		}
+	}
+	return true;
+}
+
+struct view *
+view_first(struct wl_list *head, enum lab_view_criteria criteria)
+{
+	struct view *view = wl_container_of(head->next, view, link);
+	struct server *server = view->server;
+	for (;;) {
+		if (matches_criteria(view, criteria)) {
+			break;
+		}
+		view = wl_container_of(view->link.next, view, link);
+		if (&view->link == &server->views) {
+			break;
+		}
+	}
+	return view;
+}
+
+struct view *
+view_next(struct view *view, enum lab_view_criteria criteria)
+{
+	assert(view);
+	struct server *server = view->server;
+
+	view = wl_container_of(view->link.next, view, link);
+	if (&view->link == &server->views) {
+		return view;
+	}
+	if (!matches_criteria(view, criteria)) {
+		return view_next(view, criteria);
+	}
+	return view;
+}
+
+void
+view_array_append(struct server *server, struct wl_array *views,
+		enum lab_view_criteria criteria)
+{
+	struct view *view;
+	for_each_view(&server->views, view, criteria) {
```

Comment:
I think `wl_list_for_each()` has its arguments ordered different, first the assignment arg (`view` in this case), then the list itself. Suggest ordering the arguments the same for this API to reduce confusion when using it.

Reply:
Done. Good point.